File: H_Simulator_FVM_1D/06_H_Simulator_FVM_1D_CompressibleRock/Python/src/add_wells.py
import numpy as np
from src.density import density
from src.perm import perm


def add_wells(Flow,State,Wells):

    N = len(State["P"])

    # =============================================================================
    # Add fixed-pressure wells
    # =============================================================================

    #Find permeability
    kx,dkxdP = perm(Flow,State)

    Pwf = np.zeros(N)  #Wellbore flowing pressure [Pa]
    Pwf[Wells["Loc_P"]] = Wells["P"]        #Only application location with wells

    WI = np.zeros(N)  # Well index [m^3]
    WI[Wells["Loc_P"]] = Wells["WI"]  # Only application location with wells

    Rho = np.zeros(N)  # Density [kg/m^3]
    dRhodP = np.zeros(N)  # Derivative of density wrt pressure [kg/(Pa m^3)]

    #Upstreamed density of fluid being injected
    inj = Pwf >= State["P"]
    if np.any(inj):
        Rho[inj], dRhodP[inj] = density(
            Flow,
            Pwf[inj]
        )

        # Injected fluid density does not depend on reservoir pressure
        dRhodP[inj] = 0.0

    #Upstreamed density of fluid being produced
    prod = Pwf < State["P"]
    if np.any(prod):
        Rho[prod], dRhodP[prod] = density(
            Flow,
            State["P"][prod]
        )

    #Find flow rate and derivatives
    Q = kx * (Pwf - State["P"]) * WI * Rho / Flow["muf"] #Flow rate [kg/sec]
    dQdP = kx * WI * (dRhodP * (Pwf - State["P"]) - Rho) / Flow["muf"] + dkxdP * WI * Rho / Flow["muf"]  #Derivative wrt pressure [kg/(Pa sec)]

    # =============================================================================
    # Add fixed-rate wells
    # =============================================================================

    Q_wells = np.zeros(N)
    Q_wells[Wells["Loc_Q"]] = Wells["Q"]
    #Add this contribution to source terms
    Q += Q_wells


    return Q, dQdP

File: HM_Simulator_FEM_FVM_2D/09_HM_Simulator_FEM_FVM_2D/MATLAB/src/Flow_Model/Add_Wells.m
function [Q,dQdP] = Add_Wells(Flow,State,Wells)
%% - Well Builder
%Find well source terms and derivatives wrt pressure. All upstreamed.
%Barnaby Fryer - EPFL - 29.03.17

%Calculate perm and derivative
[~,keff,~,~,dkeff_dP] = Perm(Flow,State);                       %[N,1]

%Wellbore flowing pressure [Pa]
Pwf = zeros(length(State.P),1);                                 %[N,1]  
%Only apply locations with wells
Pwf(Wells.Loc_P,1) = Wells.P;                                   %[N,1]

%Wellbore index [m]
WI = zeros(length(State.P),1);                                  %[N,1]  
%Only apply locations with wells
WI(Wells.Loc_P,1) = Wells.WI;                                   %[N,1]

%Upstreamed density of fluid being injected
[Rho(Pwf>=State.P,1),dRhodP(Pwf>=State.P,1)] =...
    Density(Flow,Pwf(Pwf>=State.P,1));                          %[N,1]
dRhodP(Pwf>=State.P,1) = 0;                                     %[N,1]
%Upstreamed density of fluid being produced
[Rho(Pwf<State.P,1),dRhodP(Pwf<State.P,1)] =...
    Density(Flow,State.P(Pwf<State.P,1));                       %[N,1]

%Flow rate [kg/sec]
Q = keff.*(Pwf - State.P).*WI.*Rho/Flow.muf;                    %[N,1]     
%Derivative of flow rate wrt pressure [kg/(Pa*sec)]
dQdP = keff.*WI.*(dRhodP.*(Pwf - State.P) - Rho)/Flow.muf ...
    + dkeff_dP.*(Pwf - State.P).*WI.*Rho/Flow.muf;              %[N,1]

%% - Add fixed rate wells

%Get fixed well data
Q_wells = zeros(length(State.P),1);                             %[N,1]  
Q_wells(Wells.Loc_Q,1) = Wells.Q;                               %[N,1]
%Add this contribution to source terms
Q = Q + Q_wells;                                                %[N,1]

end

